enhancements/ingress/optional-ingress-hypershift.md

```diff
@@ -0,0 +1,272 @@
+- HyperShift skips the deployment of the ingress operator on the hosted control plane.
+- HyperShift creates `ClusterVersion` without the ingress capability on the hosted cluster.
+- Hosted cluster's cluster version operator does not deploy the cluster ingress operator's payload (CRDs, RBACs, etc.).
+
+#### Setup AWS ALB as the new ingress
+
+- Cluster instance admin installs [AWS Load Balancer Operator](/enhancements/ingress/aws-load-balancer-operator.md).
+- Cluster instance admin sets up the default `AWSLoadBalancerController`.
+- Cluster instance user creates the `Ingress` resource for the default aws load balancer controller's ingress class.
+- AWS load balancer controller provisions an ALB on AWS.
+- Cluster instance admin can install [External DNS Operator](/enhancements/dns/externaldns-operator.md) to help cluster instance users set up convenient DNS names.
+
+### API Extensions
+
+N/A - no api extensions are being introduced in this proposal
+
+### Implementation Details/Notes/Constraints
+
+The name of the new capability is `Ingress`. It includes the cluster ingress operator and all its APIs listed below.
+The implementation of the new capability follows the instructions of [how to implement a new capability](/enhancements/installer/component-selection.md#how-to-implement-a-new-capability).
+
+#### HyperShift
+
+HyperShift doesn't pass all the ingress operator's payload to the cluster version operator.
+Certain manifests are filtered out, most importantly: the ingress operator's deployment.
+The ingress operator is considered as part of the hosted control plane, that is a resource not managed by the user hence deployed on a different cluster.
+Theoretically this may question the need for the cluster capability on HyperShift. However:
+- the cluster capability is a common interface for making OpenShift cluster components optional
+- the cluster capability is already implemented for some operators used by HyperShift (e.g. the console operator)
+- the capability provides a clear API which can be used by any party to understand whether the ingress is enabled
+- [the ingress operator's APIs](#ingress-operator-apis) are still created by the cluster version operator
+
+#### Standalone OpenShift
+
+Making the ingress operator optional for the standalone OpenShift is a bigger scope than for HyperShift.
+The changes impact the components which are not used (or used differently) in HyperShift like the installer, the authentication and monitoring operators. To prevent the scope creep for [RFE-3395](https://issues.redhat.com/browse/RFE-3395), the cluster version operator should be able to add required capabilities.
+The required capabilities should not be exposed in the capability API (install config or `ClusterVersion` object). They should be configurable in the cluster version operator and should be [implicitly enabled](https://github.com/openshift/cluster-version-operator/blob/master/pkg/cvo/status.go#L161-L165) regardless of the baseline.
```

Comment:
> They should be configurable in the cluster version operator

do you mean they should NOT be configurable in the CVO?

Reply:
> The required capabilities should not be exposed in the capability API (install config or `ClusterVersion` object).

this seems to be in conflict w/ the suggestion below that the Console operator look at whether the ingress capability is enabled to make certain decisions.  How will the Console operator make this determination if the capability is not exposed?  The cap will at least be exposed under clusterversion.status.capabilities.enabledCapabilities, even if it can't be set under clusterversion.spec.capabilities, right?

Reply:
> i like this approach (thanks @wking who i think came up with it) but we should be clear that this isn't really implementing a new capability in the cluster. It is putting a lot of the scaffolding/backing implementation in place to make it possible to disable a component in the future, but it's not really a true capability since users can't turn it off(or on), at least in standalone (i'm still unclear what users will be able to do, or how, in hypershift).
> 
> For standalone clusters, effectively it's a read-only capability.


Yes, we can say it's not a true capability. It's a first round of the implementation.

Reply:
> > They should be configurable in the cluster version operator
> 
> do you mean they should NOT be configurable in the CVO?

By "they" I meant the required capabilities. Trevor proposed a flag for the cluster version operator. So that flag will make the required capabilities configurable.

Reply:
> The cap will at least be exposed under clusterversion.status.capabilities.enabledCapabilities, even if it can't be set under clusterversion.spec.capabilities, right?

Yes. That's exactly what I wanted to say: the required capability is not a new field in `clusterversion.spec.capabilities` but it must be in `clusterversion.status.capabilities.enabledCapabilities`.
Let me rephrase this one to avoid the confusion.